File: app/lib/models/Task.dart
import 'TaskComment.dart';
import 'TaskStatus.dart';

class Task {
  int? taskId;
  String? text;
  String? subject;
  int? typeId;
  List<TaskComment>? comments = [];
  TaskStatus? status;
  int? projectId;
  int? minutes;

  Task({
    this.taskId,
    this.text,
    this.subject,
    this.status,
    this.typeId,
    this.comments,
    this.projectId,
    this.minutes,
  });

  static Task fromJson(Map<String, dynamic> json) {
    try {
      int? statusIndex = json["status"];

      return Task(
        taskId: json["taskId"],
        projectId:
            json["project"] != null ? json["project"]["projectId"] : null,
        text: json["text"],
        subject: json["subject"],
        status: statusIndex != -1 ? TaskStatus.values[statusIndex!] : TaskStatus.Archived,
        typeId: json["typeId"],
        minutes: json["minutes"],
        comments: json["comments"] != null && json["comments"].length > 0
            ? json["comments"].map<TaskComment>((t) {
                return TaskComment.fromJson(t);
              }).toList()
            : [],
      );
    } catch (e) {
      print(e);
      throw(e);
    }
  }

  setComments(List<TaskComment> comments) {
    this.comments = comments;
  }

  // USED FOR GENERIC FUNCTIONS DO NOT REMOVE
  @override
  String toString() {
    return this.text!;
  }

  void setStatus(TaskStatus status) {
    this.status = status;
  }
}

File: web/src/models/Task.ts
import { TaskComment } from './TaskComment';
import { GantChartItem } from './GantChartItem';
import { ProjectImage } from "./ProjectImage";

export const TaskStatus = {
	Backlog: 0,
	Todo: 1,
	Doing: 2,
	Done: 3,
	Archived: -1
} as const;

export type TaskStatusType = 0 | 1 | 2 | 3 | -1;

export interface Task {
	taskId: number;
	projectId: number;
	typeId: number;
	text: string;
	status: TaskStatusType;
	minutes: number; // Minutes tracked
	comments: TaskComment[];
	priority: number;
	startDate: number | null;
	endDate: number | null;
	images: ProjectImage[];
}

export class TaskItem extends GantChartItem implements Task {
	taskId: number;
	projectId: number;
	typeId: number;
	text: string;
	status: TaskStatusType;
	minutes: number; // Minutes tracked
	comments: TaskComment[];
	priority: number;
	images: ProjectImage[];

	constructor(task: Task) {
		super(task.startDate, task.endDate);
		this.taskId = task.taskId;
		this.projectId = task.projectId;
		this.typeId = task.typeId;
		this.text = task.text;
		this.status = task.status;
		this.minutes = task.minutes;
		this.comments = task.comments;
		this.priority = task.priority;
		this.images = task.images;
	}

	get taskJson(): Task {
		return {
			taskId: this.taskId,
			projectId: this.projectId,
			typeId: this.typeId,
			text: this.text,
			status: this.status,
			minutes: this.minutes,
			comments: this.comments,
			priority: this.priority,
			startDate: this.startDate,
			endDate: this.endDate,
			images: this.images
		};
	}
}
